### PR TITLE
Service Broker

### PR DESCRIPTION
**Pull request by nonemax**

### CODE REVIEW

File: broker/config.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package broker
+
+import (
+	"github.com/Percona-Lab/percona-dbaas-cli/dbaas"
+)
+
+const (
+	pxcServiceID     = "pxc-service-broker-id"
+	pxcServiceName   = "percona-xtradb-cluster"
+	psmdbServiseID   = "percona-server-for-mongodb"
```
Comment (idexter) [inline broker/config.go]:
```suggestion
	psmdbServiceID   = "percona-server-for-mongodb"
```
Hi folks. Here is the "typo" 😊. And in other places where this const has used.

File: broker/controller.go [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+					ImageURL:            "https://www.percona.com/blog/wp-content/uploads/2016/06/Percona-XtraDB-Cluster-certification-1-300x250.png",
+					ProviderDisplayName: "percona",
+				},
+				Tags: []string{
+					"pxc",
+				},
+				PlanUpdateable: true,
+			},
+			Service{
+				ID:          psmdbServiseID,
```
Comment (idexter) [inline broker/controller.go]:
```suggestion
				ID:          psmdbServiceID,
```

File: broker/cluster.go [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+						return nil
+					}
+					log.Println("Avaliable clusters:")
+					log.Print(list)
+				default:
+					log.Printf("[ERROR] create pxc: %v", err)
+				}
+			}
+		}
+	case psmdbServiseID:
```
Comment (idexter) [inline broker/cluster.go]:
```suggestion
	case psmdbServiceID:
```

File: broker/cluster.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+			select {
+			case <-ok:
+				log.Println("Deleting...[done]")
+				return nil
+			case err := <-cerr:
+				log.Printf("[ERROR] delete pxc: %v", err)
+				return err
+			}
+		}
+	case psmdbServiseID:
```
Comment (idexter) [inline broker/cluster.go]:
```suggestion
	case psmdbServiceID:
```

File: broker/cluster.go [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+	"github.com/Percona-Lab/percona-dbaas-cli/dbaas"
+	"github.com/Percona-Lab/percona-dbaas-cli/dbaas/psmdb"
+	"github.com/Percona-Lab/percona-dbaas-cli/dbaas/pxc"
+)
+
+const (
+	defaultVersion = "default"
+
+	noS3backupWarn = `[Error] S3 backup storage options doesn't set: %v. You have specify S3 storage in order to make backups.
```
Comment (dAdAbird) [inline broker/cluster.go]: this const seems to be redundant

File: broker/controller.go [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+	params.Replicas = instance.Parameters.Replicas
+	params.Size = instance.Parameters.Size
+	params.TopologyKey = instance.Parameters.TopologyKey
+
+	instanceID := ExtractVarsFromRequest(r, "service_instance_guid")
+
+	skipS3 := true
+	err = c.DeployCluster(instance, &skipS3, instanceID)
+	if err != nil {
+		log.Println("Deploy cluster", err)
```
Comment (dAdAbird) [inline broker/controller.go]: Shouldn't we abort execution and send some HTTP error in that case?

File: broker/cluster.go [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+
+		log.Println(setupmsg)
+
+		created := make(chan string)
+		msg := make(chan dbaas.OutuputMsg)
+		cerr := make(chan error)
+
+		go dbaas.Create("pxc", app, created, msg, cerr)
+
+		for {
```
Comment (dAdAbird) [inline broker/cluster.go]: All this workaround is needed for the CLI to have a user interaction (show spinner/progress, errors/messages, etc.) while the payload is executing.

File: cmd/percona-dbaas/service-broker/service-broker.go [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+	"github.com/spf13/cobra"
+)
+
+// PxcBrokerCmd represents the pxc broker command
+var PxcBrokerCmd = &cobra.Command{
+	Use:   "pxc-broker",
+	Short: "Start PXC broker",
+	Run: func(cmd *cobra.Command, args []string) {
+		log.Println("Starting broker")
+		server, err := server.NewPXCBroker(cmd.Flag("port").Value.String())
```
Comment (dAdAbird) [inline cmd/percona-dbaas/service-broker/service-broker.go]: Why only PXC?

File: broker/controller.go [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+
+	c.bindingMap[bindingID] = &ServiceBinding{
+		ID:                bindingID,
+		ServiceID:         instance.ServiceID,
+		ServicePlanID:     instance.PlanID,
+		ServiceInstanceID: instance.ID,
+	}
+
+	credentials := Credential{
+		UserName:   "PXCUser",
```
Comment (dAdAbird) [inline broker/controller.go]: PSMDB?

File: broker/controller.go [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+
+	// User - the username
+	User string
+}
+
+// New creates new controller
+func New() (Controller, error) {
+	var pxc Controller
+	pxc.instanceMap = make(map[string]*ServiceInstance)
+	pxc.bindingMap = make(map[string]*ServiceBinding)
```
Comment (dAdAbird) [inline broker/controller.go:63]: What happens with these maps on the broker restart?

File: broker/controller.go [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+	if err != nil {
+		return err
+	}
+	log.Println(string(body))
+	err = json.Unmarshal(body, object)
+	if err != nil {
+		return err
+	}
+
+	return nil
```
Comment (dAdAbird) [inline broker/controller.go:322]:
```go
return json.Unmarshal(body, object)
```

File: broker/controller.go [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+}
+
+func ProvisionDataFromRequest(r *http.Request, object interface{}) error {
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return err
+	}
+	log.Println(string(body))
```
Comment (dAdAbird) [inline broker/controller.go]: ?

File: broker/service-instance.go [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+	InternalID       string `json:"internalId,omitempty"`
+	ServiceID        string `json:"service_id"`
+	PlanID           string `json:"plan_id"`
+	OrganizationGUID string `json:"organization_guid"`
+	SpaceGUID        string `json:"space_guid"`
+
+	LastOperation *LastOperation `json:"last_operation,omitempty"`
+
+	Parameters struct {
+		PXCParameters
```
Comment (dAdAbird) [inline broker/service-instance.go]: PSMDB parameters?

File: broker/config.go [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+	c.PXC.StorageClass = ""
+	c.PXC.Instances = int32(3)
+	c.PXC.RequestCPU = "600m"
+	c.PXC.RequestMem = "1G"
+	c.PXC.AntiAffinityKey = "kubernetes.io/hostname"
+	c.ProxySQL.Instances = int32(1)
+	c.ProxySQL.RequestCPU = "600m"
+	c.ProxySQL.RequestMem = "1G"
+	c.ProxySQL.AntiAffinityKey = "kubernetes.io/hostname"
+	c.S3.SkipStorage = true
```
Comment (dAdAbird) [inline broker/config.go:36]: I don't like the idea of mixin up PXC and PSMD parameters

File: broker/controller.go [hunk 15]
```diff
@@ -0,0 +1,328 @@
...
+	log.Println("Remove Service Instance...")
+
+	instanceID := ExtractVarsFromRequest(r, "service_instance_guid")
+	instance := c.instanceMap[instanceID]
+	if instance == nil {
+		w.WriteHeader(http.StatusGone)
+		return
+	}
+
+	c.DeletePXCCluster(instance)
```
Comment (dAdAbird) [inline broker/controller.go]: error handling

File: broker/cluster.go [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+			}
+		}
+	}
+	return nil
+}
+
+func (p *Controller) DeletePXCCluster(instance *ServiceInstance) error {
+	ok := make(chan string)
+	cerr := make(chan error)
+	delePVC := true
```
Comment (dAdAbird) [inline broker/cluster.go]: Should we delete PVC? Should it be an option (can it be)?

File: broker/cluster.go [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+func (p *Controller) DeployCluster(instance ServiceInstance, skipS3Storage *bool, instanceID string) error {
+	switch instance.ServiceID {
+	case pxcServiceID:
+		app := pxc.New(instance.Parameters.ClusterName, defaultVersion)
+		conf := dbaas.ClusterConfig{}
+		SetDefault(&conf)
+
+		var s3stor *dbaas.BackupStorageSpec
+
+		setupmsg, err := app.Setup(conf, s3stor)
```
Comment (dAdAbird) [inline broker/cluster.go:33]: Check if there is no panic later on

File: broker/config.go (hunk 1 above)
Comment (nonemax) [reply broker/config.go]: Thanks, will fix

File: broker/cluster.go (hunk 7 above)
Comment (dAdAbird) [inline broker/cluster.go]: All this workaround is needed for the CLI to have a user interaction (show spinner/progress, errors/messages, etc.) while the payload is executing.

**Review by dAdAbird — CHANGES_REQUESTED** [review]

File: broker/cluster.go (hunk 16 above)
Comment (nonemax) [reply broker/cluster.go]: set to "true"

File: broker/cluster.go (hunk 17 above)
Comment (nonemax) [reply broker/cluster.go:33]: no panics wile testing

File: broker/cluster.go [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+func (p *Controller) DeleteCluster(instance *ServiceInstance) error {
+	ok := make(chan string)
+	cerr := make(chan error)
+	delePVC := false
+	name := instance.Parameters.ClusterName
+	switch instance.ServiceID {
+	case pxcServiceID:
+		go dbaas.Delete("pxc", pxc.New(name, defaultVersion), delePVC, ok, cerr)
+		tckr := time.NewTicker(1 * time.Second)
+		defer tckr.Stop()
```
Comment (dAdAbird) [inline broker/cluster.go]: the redundant ticker
the same bellow

**Review by dAdAbird — APPROVED** [review]

**Review by idexter — APPROVED** [review]

**Review by delgod — APPROVED** [review]

**Pull request merged**